ui/StatusQ/src/singleModelItemData.cpp

```diff
@@ -0,0 +1,254 @@
+#include "StatusQ/SingleModelItemData/singleModelItemData.h"
+#include <QDebug>
+
+SingleModelItemData::SingleModelItemData(QObject *parent)
+    : QObject(parent)
+    , m_item(new QQmlPropertyMap(this))
+{
+}
+
+QQmlPropertyMap* SingleModelItemData::item() const
+{
+    return m_item.data();
+}
+
+QAbstractItemModel* SingleModelItemData::sourceModel() const
+{
+    return m_sourceModel.data();
+}
+
+SingleItemFilter* SingleModelItemData::filter() const
+{
+    return m_filter.data();
+}
+
+bool SingleModelItemData::available() const
+{
+    return m_available;
+}
+
+const QStringList& SingleModelItemData::roles() const
+{
+    return m_roles;
+}
+
+int SingleModelItemData::row() const
+{
+    return m_index.row();
+}
+
+void SingleModelItemData::setSourceModel(QAbstractItemModel* sourceModel)
+{
+    if (m_sourceModel == sourceModel)
+        return;
+
+    if (m_sourceModel)
+    {
+        disconnect(m_sourceModel, nullptr, this, nullptr);
+    }
+    m_sourceModel = sourceModel;
+
+    resetIndex();
+
+    if (!m_sourceModel)
+    {
+        emit sourceModelChanged();
+        return;
+    }
+    
+    connect(m_sourceModel, &QAbstractItemModel::modelReset, this, [this]() {
+        resetIndex();
+    });
+    connect(m_sourceModel, &QAbstractItemModel::rowsInserted, this, [this]() {
+        if (!m_index.isValid())
+        {
+            resetIndex();
+        }
+    });
+    connect(m_sourceModel, &QAbstractItemModel::rowsMoved, this, [this](const QModelIndex &parent, int start, int end, const QModelIndex &destination, int row) {
+        if (!m_index.isValid())
+            return;
+
+        if (m_index.row() >= destination.row() && m_index.row() <= destination.row() + (end - start))
+        {
+            emit rowChanged();
+        }
+    });
+    connect(m_sourceModel, &QAbstractItemModel::rowsAboutToBeRemoved, this, [this](const QModelIndex &parent, int first, int last) {
+        if (!m_index.isValid())
+            return;
+        
+        if (m_index.row() >= first && m_index.row() <= last)
+        {
+            setIndex({});
+        }
+    });
+    connect(m_sourceModel, &QAbstractItemModel::dataChanged, this, [this](const QModelIndex &topLeft, const QModelIndex &bottomRight, const QList<int> &roles = QList<int>()) {        
+        if (!m_index.isValid())
+        {
+            auto index = findIndexInRange(topLeft.row(), bottomRight.row() + 1);
+            setIndex(index);
+            return;
+        }
+        
+        // Check if the index is still valid
+        auto index = findIndexInRange(m_index.row(), m_index.row() + 1);
```

Comment:
This lookup on every data change in practice will be not necessary in vast majority of cases. The values of role treated as a key will never change or in very rare cases. Because of that I wonder if the `SingleItemFilter` is justified as it blocks us from skipping data changes on irrelevant roles. Or maybe filter should return roles that it takes into account? :thinking:

Reply:
good point! We shouldn't do an unnecessary search. We don't have to implement the `SingleItemFilter` now. We can provide now the optimised `SingleModelItemData` and it can be added later if needed.
I'll update.